File: src/binary_view.c
#include "binary_view.h"
#include <string.h>
#ifndef _WIN32
#include <strings.h>
#endif

static const char* k_data_sections[] = {
    ".data", ".data1", ".rdata", ".idata", ".edata", ".reloc", ".tls", NULL
};

int bv_is_data_section_name(const char* name) {
    for (int i = 0; k_data_sections[i]; i++)
#ifdef _WIN32
        if (_stricmp(name, k_data_sections[i]) == 0) return 1;
#else
        if (strcasecmp(name, k_data_sections[i]) == 0) return 1;
#endif
    return 0;
}

void bv_from_xbe(const Xbe* xbe, BinaryView* out) {
    memset(out, 0, sizeof *out);
    out->base_address = xbe->base_address;
    out->image_size = xbe->image_size;
    out->entry_point = xbe->entry_point;
    out->kernel_thunk_table_start = xbe->kernel_thunk_addr;

    for (size_t i = 0; i < xbe->sections.len; i++) {
        const XbeSection* s = &xbe->sections.data[i];
        SectionView sv;
        memset(&sv, 0, sizeof sv);
        snprintf(sv.name, sizeof sv.name, "%s", s->name);
        sv.base_address = s->virtual_addr;
        sv.size = s->virtual_size;
        sv.executable = (s->flags & XBE_SEC_EXECUTABLE) != 0;
        sv.writable = (s->flags & XBE_SEC_WRITABLE) != 0;
        sv.data = calloc(1, s->virtual_size ? s->virtual_size : 1);
        uint32_t copy = s->raw_size < s->virtual_size ? s->raw_size : s->virtual_size;
        if (copy && (size_t)(s->raw_addr) + copy <= xbe->raw_len)
            memcpy(sv.data, xbe->raw + s->raw_addr, copy);
        vec_push(&out->sections, sv);
    }

    for (size_t i = 0; i < xbe->kernel_imports.len; i++) {
        const XbeKernelImport* k = &xbe->kernel_imports.data[i];
        ImportSymbol sym;
        memset(&sym, 0, sizeof sym);
        sym.address = k->thunk_addr;
        snprintf(sym.name, sizeof sym.name, "xboxkrnl@%d:%s", k->ordinal, k->name);
        vec_push(&out->imports, sym);
    }
}

void bv_free(BinaryView* v) {
    for (size_t i = 0; i < v->sections.len; i++) free(v->sections.data[i].data);
    vec_free(&v->sections);
    vec_free(&v->imports);
    memset(v, 0, sizeof *v);
}

SectionView* bv_find_section(BinaryView* v, uint32_t addr) {
    for (size_t i = 0; i < v->sections.len; i++) {
        SectionView* s = &v->sections.data[i];
        if (addr >= s->base_address && addr < s->base_address + s->size) return s;
    }
    return NULL;
}

SectionView* bv_find_section_by_name(BinaryView* v, const char* name) {
    for (size_t i = 0; i < v->sections.len; i++)
#ifdef _WIN32
        if (_stricmp(v->sections.data[i].name, name) == 0) return &v->sections.data[i];
#else
        if (strcasecmp(v->sections.data[i].name, name) == 0) return &v->sections.data[i];
#endif
    return NULL;
}

const uint8_t* bv_translate(BinaryView* v, uint32_t addr, size_t* avail) {
    SectionView* s = bv_find_section(v, addr);
    if (!s) { if (avail) *avail = 0; return NULL; }
    uint32_t off = addr - s->base_address;
    if (avail) *avail = s->size - off;
    return s->data + off;
}

int bv_read_u32(BinaryView* v, uint32_t addr, uint32_t* out) {
    size_t avail;
    const uint8_t* p = bv_translate(v, addr, &avail);
    if (!p || avail < 4) return 0;
    *out = (uint32_t)p[0] | ((uint32_t)p[1] << 8) | ((uint32_t)p[2] << 16) | ((uint32_t)p[3] << 24);
    return 1;
}

int bv_is_executable(BinaryView* v, uint32_t addr) {
    SectionView* s = bv_find_section(v, addr);
    return s && s->executable;
}

File: src/OgXbox.Recomp.Codegen/Binary/BinaryView.cs
// BinaryView — self-contained, VA-addressable view of a loaded binary that owns
// its section bytes. The codegen phases only ever see this, never the raw XBE.
//
// Ported to C# / x86 from rex::codegen::BinaryView (ReXGlue SDK, BSD-3-Clause).
// ReXGlue's is built from a XEX Module; this one is built from an Xbe.

using System;
using System.Collections.Generic;
using System.Linq;

namespace OgXbox.Recomp.Codegen.Binary;

/// <summary>A VA range backed by owned bytes.</summary>
public sealed class SectionView
{
    public required string Name { get; init; }
    public required uint BaseAddress { get; init; }
    public required uint Size { get; init; }
    public required byte[] Data { get; init; }   // length == Size; zero-padded past RawSize (BSS tail)
    public required bool Executable { get; init; }
    public required bool Writable { get; init; }

    public uint End => BaseAddress + Size;
    public bool Contains(uint addr) => addr >= BaseAddress && addr < End;

    public ReadOnlySpan<byte> Translate(uint addr) =>
        Contains(addr) ? Data.AsSpan((int)(addr - BaseAddress)) : default;
}

/// <summary>An import thunk: the address a <c>call</c> targets, plus its name.</summary>
public readonly record struct ImportSymbol(uint Address, string Name);

public sealed class BinaryView
{
    // Data sections stay data even when the XBE marks them executable: Xbox
    // linkers flag almost everything executable, and disassembling zero-fill
    // produces phantom functions. XDK library sections (D3D_RD etc.) are real
    // code and are NOT excluded.
    private static readonly HashSet<string> DataSectionNames = new(StringComparer.OrdinalIgnoreCase)
    {
        ".data", ".data1", ".rdata", ".idata", ".edata", ".reloc", ".tls",
    };

    private readonly List<SectionView> _sections;
    private readonly List<ImportSymbol> _imports;

    private BinaryView(List<SectionView> sections, List<ImportSymbol> imports,
                       uint baseAddress, uint imageSize, uint entryPoint, uint kernelThunkAddr)
    {
        _sections = sections;
        _imports = imports;
        BaseAddress = baseAddress;
        ImageSize = imageSize;
        EntryPoint = entryPoint;
        KernelThunkTableStart = kernelThunkAddr;
    }

    public uint BaseAddress { get; }
    public uint ImageSize { get; }
    public uint EntryPoint { get; }

    /// <summary>Start of the kernel import thunk table (0 if none). Not code.</summary>
    public uint KernelThunkTableStart { get; }

    public IReadOnlyList<SectionView> Sections => _sections;
    public IReadOnlyList<ImportSymbol> ImportSymbols => _imports;

    public SectionView? FindSection(uint addr) => _sections.FirstOrDefault(s => s.Contains(addr));

    public SectionView? FindSectionByName(string name) =>
        _sections.FirstOrDefault(s => string.Equals(s.Name, name, StringComparison.OrdinalIgnoreCase));

    public ReadOnlySpan<byte> Translate(uint addr)
    {
        var sec = FindSection(addr);
        return sec is null ? default : sec.Translate(addr);
    }

    public bool IsExecutable(uint addr) => FindSection(addr) is { Executable: true };

    public uint? ReadU32(uint addr)
    {
        var span = Translate(addr);
        if (span.Length < 4) return null;
        return System.Buffers.Binary.BinaryPrimitives.ReadUInt32LittleEndian(span);
    }

    /// <summary>
    /// Sections suitable for disassembly: executable, non-empty, and not a
    /// conventional data section by name.
    /// </summary>
    public IEnumerable<SectionView> CodeSections =>
        _sections.Where(s => s.Executable && s.Size > 0 && !DataSectionNames.Contains(s.Name));

    /// <summary>Construct directly from section views (tests, synthetic inputs).</summary>
    internal static BinaryView FromSections(
        IEnumerable<SectionView> sections,
        uint baseAddress = 0x00010000, uint entryPoint = 0, uint kernelThunkAddr = 0,
        IEnumerable<ImportSymbol>? imports = null)
    {
        var secList = sections.ToList();
        uint imageSize = secList.Count == 0 ? 0
            : secList.Max(s => s.End) - baseAddress;
        return new BinaryView(secList, (imports ?? Enumerable.Empty<ImportSymbol>()).ToList(),
                              baseAddress, imageSize, entryPoint, kernelThunkAddr);
    }

    public static BinaryView FromXbe(Xbe xbe)
    {
        var sections = new List<SectionView>();
        foreach (var s in xbe.Sections)
        {
            // Own a VirtualSize buffer; copy RawSize bytes, leave the rest zero (BSS tail).
            var buf = new byte[s.VirtualSize];
            int copy = (int)Math.Min(s.RawSize, s.VirtualSize);
            if (copy > 0 && s.RawAddress + copy <= xbe.RawData.Length)
                Array.Copy(xbe.RawData, s.RawAddress, buf, 0, copy);

            sections.Add(new SectionView
            {
                Name = s.Name,
                BaseAddress = s.VirtualAddress,
                Size = s.VirtualSize,
                Data = buf,
                Executable = s.IsExecutable,
                Writable = s.IsWritable,
            });
        }

        var imports = xbe.KernelImports
            .Select(k => new ImportSymbol(k.ThunkAddress, $"xboxkrnl@{k.Ordinal}:{k.Name}"))
            .ToList();

        return new BinaryView(sections, imports, xbe.BaseAddress, xbe.ImageSize,
                              xbe.EntryPoint, xbe.KernelThunkAddress);
    }
}
